src/Features/LanguageServer/Protocol/Handler/Configuration/DidChangeConfigurationNotificationHandler.cs

```diff
@@ -0,0 +1,212 @@
+    {
+        private readonly ILspLogger _lspLogger;
+        private readonly IGlobalOptionService _globalOptionService;
+        private readonly IClientLanguageServerManager _clientLanguageServerManager;
+        private readonly IAsynchronousOperationListener _asynchronousOperationListener;
+        private readonly Guid _registrationId;
+
+        /// <summary>
+        /// All the global <see cref="ConfigurationItem.Section"/> needs to be refreshed from the client. 
+        /// </summary>
+        private readonly ImmutableArray<ConfigurationItem> _configurationItems;
+        private static readonly ImmutableDictionary<string, string> s_languageNameToPrefix = ImmutableDictionary<string, string>.Empty
+            .Add(LanguageNames.CSharp, "csharp")
+            .Add(LanguageNames.VisualBasic, "visual_basic");
+
+        public static readonly ImmutableArray<string> SupportedLanguages = ImmutableArray.Create(LanguageNames.CSharp, LanguageNames.VisualBasic);
+
+        public DidChangeConfigurationNotificationHandler(
+            ILspLogger logger,
+            IGlobalOptionService globalOptionService,
+            IClientLanguageServerManager clientLanguageServerManager,
+            IAsynchronousOperationListener asynchronousOperationListener)
+        {
+            _lspLogger = logger;
+            _globalOptionService = globalOptionService;
+            _clientLanguageServerManager = clientLanguageServerManager;
+            _registrationId = Guid.NewGuid();
+            _asynchronousOperationListener = asynchronousOperationListener;
+            _configurationItems = GenerateGlobalConfigurationItems();
+        }
+
+        public bool MutatesSolutionState => true;
+
+        public bool RequiresLSPSolution => false;
+
+        public async Task HandleNotificationAsync(DidChangeConfigurationParams request, RequestContext requestContext, CancellationToken cancellationToken)
+        {
+            using var _ = _asynchronousOperationListener.BeginAsyncOperation(Methods.WorkspaceDidChangeConfigurationName);
+            await RefreshOptionsAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task RefreshOptionsAsync(CancellationToken cancellationToken)
+        {
+            var configurationsFromClient = await GetConfigurationsAsync(cancellationToken).ConfigureAwait(false);
+            if (configurationsFromClient.IsEmpty)
+            {
+                // Failed to get values from client, do nothing.
+                return;
+            }
+
+            // We always fetch VB and C# value from client if the option is IPerLanguageValuedOption.
+            RoslynDebug.Assert(configurationsFromClient.Length == SupportedOptions.Sum(option => option is IPerLanguageValuedOption ? 2 : 1));
+            var optionsToRefresh = SupportedOptions.SelectManyAsArray(option => option is IPerLanguageValuedOption
```

Comment:
This looks like it doesn't depend on the current request, correct?  If so could be made into a static field